File: frontend/scripts/crap/formatter.ts
/**
 * Report formatter. Mirrors `examples/crap4java/src/crap4java/ReportFormatter.java`.
 * Renders three views:
 *   - per-file summary (max CRAP, function count over threshold)
 *   - top-N worst offenders (sorted by CRAP descending)
 *   - overall histogram + distribution
 */

import type { FunctionMetric } from "./analyzer";

export interface FormatOptions {
	threshold: number;
	topN: number;
}

export function formatReport(metrics: FunctionMetric[], options: FormatOptions): string {
	const lines: string[] = [];
	const total = metrics.length;
	const withCoverage = metrics.filter((m) => m.crap !== null);
	const overThreshold = withCoverage.filter((m) => (m.crap ?? 0) >= options.threshold);

	lines.push(`# CRAP Analysis (threshold ≥ ${options.threshold})`);
	lines.push("");
	lines.push(`Functions analyzed: ${total}`);
	lines.push(`Functions with coverage data: ${withCoverage.length}`);
	lines.push(`Functions over threshold: ${overThreshold.length}`);
	lines.push("");

	lines.push(`## Top ${options.topN} worst CRAP scores`);
	lines.push("");
	const top = withCoverage.toSorted((a, b) => (b.crap ?? 0) - (a.crap ?? 0)).slice(0, options.topN);
	if (top.length === 0) {
		lines.push("(no functions with coverage data)");
	} else {
		const w = {
			crap: 7,
			cc: 4,
			cov: 6,
			loc: 8,
		};
		lines.push(
			`${pad("CRAP", w.crap)}  ${pad("CC", w.cc)}  ${pad("Cov%", w.cov)}  ${pad("Lines", w.loc)}  Function`
		);
		lines.push("-".repeat(80));
		for (const m of top) {
			const cov = m.coverage === null ? "-" : (m.coverage * 100).toFixed(1);
			lines.push(
				[
					pad((m.crap ?? 0).toFixed(2), w.crap),
					pad(String(m.complexity), w.cc),
					pad(cov, w.cov),
					pad(`${m.startLine}-${m.endLine}`, w.loc),
					`${m.file} :: ${m.name}`,
				].join("  ")
			);
		}
	}

	lines.push("");
	lines.push("## Distribution");
	lines.push("");
	const buckets = [
		{ label: "CRAP < 4 (clean)", lo: 0, hi: 4 },
		{ label: "4 ≤ CRAP < 8", lo: 4, hi: 8 },
		{ label: "8 ≤ CRAP < 30", lo: 8, hi: 30 },
		{ label: "CRAP ≥ 30 (crisis)", lo: 30, hi: Number.POSITIVE_INFINITY },
	];
	for (const b of buckets) {
		const count = withCoverage.filter((m) => {
			const c = m.crap ?? 0;
			return c >= b.lo && c < b.hi;
		}).length;
		const pct = withCoverage.length > 0 ? (100 * count) / withCoverage.length : 0;
		lines.push(`  ${pad(b.label, 22)} ${pad(String(count), 5)}  ${pct.toFixed(1)}%`);
	}

	lines.push("");
	lines.push("## Functions without coverage data");
	const uncovered = metrics.filter((m) => m.coverage === null);
	lines.push(
		`  ${uncovered.length} functions (no executable LCOV lines in span — likely type-only or unreachable)`
	);

	return lines.join("\n");
}

function pad(s: string, width: number): string {
	return s.length >= width ? s : s + " ".repeat(width - s.length);
}

File: examples/winstt-electron/server/scripts/crap/formatter.py
"""Report formatter. Mirrors `frontend/scripts/crap/formatter.ts`.

Renders: counts header, top-N worst offenders, distribution histogram,
and the count of functions with no coverage data.
"""

from __future__ import annotations

from .analyzer import FunctionMetric


def _pad(s: str, width: int) -> str:
    return s if len(s) >= width else s + " " * (width - len(s))


def format_report(metrics: list[FunctionMetric], *, threshold: float, top_n: int) -> str:
    lines: list[str] = []
    total = len(metrics)
    with_cov = [m for m in metrics if m.crap is not None]
    over = [m for m in with_cov if (m.crap or 0.0) >= threshold]

    lines.append(f"# CRAP Analysis (threshold >= {threshold})")
    lines.append("")
    lines.append(f"Functions analyzed: {total}")
    lines.append(f"Functions with coverage data: {len(with_cov)}")
    lines.append(f"Functions over threshold: {len(over)}")
    lines.append("")

    lines.append(f"## Top {top_n} worst CRAP scores")
    lines.append("")
    top = sorted(with_cov, key=lambda m: m.crap or 0.0, reverse=True)[:top_n]
    if not top:
        lines.append("(no functions with coverage data)")
    else:
        w = {"crap": 8, "cc": 4, "cov": 6, "loc": 9}
        lines.append(
            f"{_pad('CRAP', w['crap'])}  {_pad('CC', w['cc'])}  "
            f"{_pad('Cov%', w['cov'])}  {_pad('Lines', w['loc'])}  Function"
        )
        lines.append("-" * 80)
        for m in top:
            cov = "-" if m.coverage is None else f"{m.coverage * 100:.1f}"
            lines.append(
                "  ".join(
                    [
                        _pad(f"{m.crap or 0.0:.2f}", w["crap"]),
                        _pad(str(m.complexity), w["cc"]),
                        _pad(cov, w["cov"]),
                        _pad(f"{m.start_line}-{m.end_line}", w["loc"]),
                        f"{m.file} :: {m.name}",
                    ]
                )
            )

    lines.append("")
    lines.append("## Distribution")
    lines.append("")
    buckets = [
        ("CRAP < 4 (clean)", 0.0, 4.0),
        ("4 <= CRAP < 8", 4.0, 8.0),
        ("8 <= CRAP < 30", 8.0, 30.0),
        ("CRAP >= 30 (crisis)", 30.0, float("inf")),
    ]
    for label, lo, hi in buckets:
        count = sum(1 for m in with_cov if lo <= (m.crap or 0.0) < hi)
        pct = (100.0 * count / len(with_cov)) if with_cov else 0.0
        lines.append(f"  {_pad(label, 22)} {_pad(str(count), 5)}  {pct:.1f}%")

    lines.append("")
    lines.append("## Functions without coverage data")
    uncovered = [m for m in metrics if m.coverage is None]
    lines.append(
        f"  {len(uncovered)} functions (no executable coverage lines in span "
        "- likely abstract, type-only, or unreachable)"
    )

    return "\n".join(lines)
